packages/simcore-sdk/tests/unit/test_node_ports_v2_port_validation.py

```diff
@@ -0,0 +1,224 @@
+    value, unit = validate_port_content(
+        "port_1",
+        value=3.0,
+        unit="milli-meter",
+        # expected
+        content_schema={
+            "title": "simple number",
+            "type": "number",
+            "x_unit": "cm",
+        },
+    )
+
+    #  3.0 mm -> 0.3 cm
+    # NOTE: there are roundoff errors in the unit conversion
+    assert unit == "centimeter"
+    assert abs(value - 0.3) < 1e-14
+
+
+def test_validate_port_content_fails():
+
+    with pytest.raises(PortUnitError) as err_info:
+        value, unit = validate_port_content(
+            "port_1",
+            value=3.0,
+            unit="seconds",
+            # expected
+            content_schema={
+                "title": "simple number",
+                "type": "number",
+                "x_unit": "cm",
+            },
+        )
+
+    error_message = f"{err_info.value}"
+    # "Invalid unit in port 'port_1': Cannot convert from 'second' ([time]) to 'centimeter' ([length])"
+    assert "'port_1'" in error_message
+    assert "'second'" in error_message
+    assert "'centimeter'" in error_message
+
+
+async def test_port_with_array(mocker):
+    mocker.patch.object(Port, "_node_ports", new=AsyncMock())
+
+    # arrays
+    port_meta = {
+        "label": "array_numbers",
+        "description": "Some array of numbers",
+        "type": "ref_contentSchema",
+        "contentSchema": {
+            "title": "list[number]",
+            "type": "array",
+            "items": {"type": "number"},
+        },
+    }
+    expected_value = [1, 2, 3]
+
+    print(json.dumps(port_meta, indent=1))
+    print(json.dumps(expected_value, indent=1))
+
+    port = Port(key="input_w_array", **port_meta)
+
+    await port.set_value(expected_value)
+    assert await port.get_value() == expected_value
+
+
+async def test_port_with_array_of_object(mocker):
+    mocker.patch.object(Port, "_node_ports", new=AsyncMock())
+
+    class A(BaseModel):
+        i: conint(gt=3)
+        b: bool = False
+        s: str
+        l: List[int]
+
+    content_schema = schema_of(List[A], title="array[A]")
+
+    port_meta = {
+        "label": "array_",
+        "description": "Some array of As",
+        "type": "ref_contentSchema",
+        "contentSchema": content_schema,
+    }
+    sample = [{"i": 5, "s": "x", "l": [1, 2]}, {"i": 6, "s": "y", "l": [2]}]
+    expected_value = [A(**i).dict() for i in sample]
+
+    print(json.dumps(port_meta, indent=1))
+    print(json.dumps(expected_value, indent=1))
+
+    # valid data and should assign defaults
+    value = deepcopy(sample)
+    p = Port(key="k", value=value, **port_meta)
+    assert p.value == expected_value
+
+    value = deepcopy(sample)
+    value[0]["i"] = 0  # violates >3 condition
+
+    with pytest.raises(ValidationError) as excinfo:
+        Port(key="k", value=value, **port_meta)
+
+    assert (
+        "0 is less than or equal to the minimum of 3"
+        in excinfo.value.errors()[0]["msg"]
```

Comment:
So here you only expect 1 error right? Could you also add a check for the size of the errors to be 1?